applications/wikipedia/main.py

```diff
@@ -58,86 +68,208 @@ def download_model():
+                "--query-gpu=utilization.gpu",
+                "--format=csv,noheader,nounits",
+            ],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        out_str = sp.communicate()
+        out_list = out_str[0].decode("utf-8").split("\n")
+        out_list = [x for x in out_list if x]
+        return [round(float(x) / 100, 3) for x in out_list]
+    except Exception as e:
+        print("Exception: ", e)
+        return []
+
+
 @stub.cls(
     gpu=GPU_CONFIG,
     image=tei_image,
-    # Use up to 20 GPU containers at once.
-    concurrency_limit=10,
+    # Use up to 10 GPU containers at once.
+    concurrency_limit=N_GPU,
     # Allow each container to process up to 10 batches at once.
-    allow_concurrent_inputs=10,
+    allow_concurrent_inputs=100,
+    secret=Secret.from_name("wandb"),
 )
 class TextEmbeddingsInference:
     def __enter__(self):
+        # If the process is running for a long time, the client does not seem to close the connections, results in a pool timeout
         from httpx import AsyncClient
+        import wandb
 
+        self.keep_running = True
         self.process = spawn_server()
         self.client = AsyncClient(base_url="http://127.0.0.1:8000")
 
+        if ENABLE_WANDB:
+            wandb.init(project=WANDB_PROJECT, group=WANDB_GROUP, reinit=True)
+
+        self.gpu_utilization_figures = []
+
+        def record_gpu_utilization():
+            while self.keep_running:
+                curr_utilization = get_gpu_utilization()
+                self.gpu_utilization_figures.append(curr_utilization[0])
+                time.sleep(1)
+
+        self.gpu_utilization_thread = threading.Thread(target=record_gpu_utilization)
+        self.gpu_utilization_thread.start()
+
     def __exit__(self, _exc_type, _exc_value, _traceback):
+        import wandb
+
+        self.keep_running = False
+        self.gpu_utilization_thread.join()
+        wandb.finish()
         self.process.terminate()
 
     @method()
-    async def embed(self, inputs: list[str]):
-        print(f"Recieved {len(inputs)} inputs")
-        resp = self.client.post("/embed", json={"inputs": inputs})
-        resp = await resp
-        resp.raise_for_status()
-        outputs = resp.json()
-
-        # Returning a list is slower because of additional Modal-specific overhead,
-        # to be fixed shortly.
-        return np.array(outputs)
+    async def embed(self, texts: list[str]):
+        start = time.perf_counter()
+        n_chars = sum(map(len, texts))
+        res = await self.client.post("/embed", json={"inputs": texts})
+        embeddings = res.json()
+        end = time.perf_counter()
+        total_time = int(end - start)
+        snapshot = self.gpu_utilization_figures[-total_time:]
+        if not snapshot:
+            snapshot = [get_gpu_utilization()[0]]
+        gpu_utilization = sum(snapshot) / len(snapshot)
+        return list(zip(texts, embeddings)), n_chars, gpu_utilization
```

Comment:
zipping is extra overhead
texts exists in the main computer so zipping and sending it back is extra overhead.